File: src/backend/infrastructure/gemini/GeminiService.ts
import { GoogleGenerativeAI, Part } from '@google/generative-ai';
import {
  IImageGenerationService,
  ImageGenerationRequest,
  ImageGenerationResponse,
  ILoggerService,
} from '../../domain/interfaces/index';
import { GeminiError } from '../../application/errors/AppErrors';

export class GeminiService implements IImageGenerationService {
  private clients: Map<string, GoogleGenerativeAI> = new Map();

  constructor(private readonly logger: ILoggerService) {}

  private getClient(apiKey: string): GoogleGenerativeAI {
    if (!this.clients.has(apiKey)) {
      this.clients.set(apiKey, new GoogleGenerativeAI(apiKey));
    }
    return this.clients.get(apiKey)!;
  }

  async generateImage(request: ImageGenerationRequest): Promise<ImageGenerationResponse> {
    const { imageBuffer, mimeType, prompt, model, quality } = request;

    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) {
      throw new GeminiError('Gemini API key not configured. Please set it in Settings.');
    }

    this.logger.info('Starting Gemini image generation', {
      model,
      mimeType,
      bufferSize: imageBuffer.length,
      promptLength: prompt.length,
    });

    try {
      const client = this.getClient(apiKey);
      const genModel = client.getGenerativeModel({
        model: model || 'gemini-2.0-flash-exp',
        generationConfig: {
          // @ts-expect-error - responseModalities is new
          responseModalities: ['Text', 'Image'],
        },
      });

      const base64Image = imageBuffer.toString('base64');
      const imagePart: Part = {
        inlineData: {
          data: base64Image,
          mimeType: mimeType as 'image/jpeg' | 'image/png' | 'image/webp',
        },
      };

      const textPart: Part = { text: prompt };

      this.logger.debug('Sending request to Gemini API', { model, quality });

      const result = await genModel.generateContent([textPart, imagePart]);
      const response = result.response;

      let resultImageBuffer: Buffer | null = null;
      let resultMimeType = 'image/png';

      for (const candidate of response.candidates || []) {
        for (const part of candidate.content?.parts || []) {
          if (part.inlineData?.mimeType?.startsWith('image/')) {
            resultImageBuffer = Buffer.from(part.inlineData.data, 'base64');
            resultMimeType = part.inlineData.mimeType;
            break;
          }
        }
        if (resultImageBuffer) break;
      }

      if (!resultImageBuffer) {
        const textResponse = response.text();
        this.logger.warn('No image in Gemini response', { textResponse });
        throw new GeminiError(
          `Gemini did not return an image. Response: ${textResponse?.slice(0, 200) || 'empty'}`
        );
      }

      const tokensUsed = response.usageMetadata?.totalTokenCount;

      this.logger.info('Gemini image generation successful', {
        outputSize: resultImageBuffer.length,
        outputMimeType: resultMimeType,
        tokensUsed,
      });

      return { imageBuffer: resultImageBuffer, mimeType: resultMimeType, tokensUsed };
    } catch (error) {
      if (error instanceof GeminiError) throw error;
      const err = error as Error;
      this.logger.error('Gemini API error', err);

      if (err.message?.includes('API_KEY_INVALID') || err.message?.includes('400')) {
        throw new GeminiError('Invalid Gemini API key. Please check your settings.');
      }
      if (err.message?.includes('RATE_LIMIT') || err.message?.includes('429')) {
        throw new GeminiError('Gemini rate limit exceeded. Please wait and retry.');
      }
      if (err.message?.includes('SAFETY') || err.message?.includes('blocked')) {
        throw new GeminiError('Image was blocked by Gemini safety filters.');
      }
      if (err.message?.includes('quota') || err.message?.includes('QUOTA')) {
        throw new GeminiError('Gemini API quota exceeded for today.');
      }

      throw new GeminiError(`Gemini API error: ${err.message}`);
    }
  }

  async validateCredentials(): Promise<boolean> {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) return false;
    return this.validateApiKey(apiKey);
  }

  async validateApiKey(apiKey: string): Promise<boolean> {
    try {
      const client = this.getClient(apiKey);
      const model = client.getGenerativeModel({ model: 'gemini-1.5-flash' });
      const result = await model.generateContent('Say "ok" in one word.');
      const text = result.response.text();
      return text.length > 0;
    } catch {
      return false;
    }
  }

  updateApiKey(newApiKey: string): void {
    process.env.GEMINI_API_KEY = newApiKey;
    this.clients.delete(newApiKey);
  }
}

File: app/infrastructure/providers/gemini_service.py
"""Gemini image-generation provider.

Python equivalent of `src/backend/infrastructure/gemini/GeminiService.ts`.
Uses the official `google-generativeai` SDK. Blocking SDK calls are offloaded
to a thread so the asyncio event loop is never blocked.
"""
from __future__ import annotations

import asyncio
import base64
import os
from typing import Optional

from app.core.errors import GeminiError
from app.core.logging import Logger
from app.domain.interfaces import (
    IImageGenerationService,
    ImageGenerationRequest,
    ImageGenerationResponse,
)


class GeminiService(IImageGenerationService):
    def __init__(self, logger: Logger) -> None:
        self._logger = logger

    async def generate_image(self, request: ImageGenerationRequest) -> ImageGenerationResponse:
        api_key = os.environ.get("GEMINI_API_KEY")
        if not api_key:
            raise GeminiError("Gemini API key not configured. Please set it in Settings.")

        self._logger.info("Starting Gemini image generation", {
            "model": request.model, "mimeType": request.mime_type,
            "bufferSize": len(request.image_buffer), "promptLength": len(request.prompt),
        })

        try:
            return await asyncio.to_thread(self._generate_sync, request, api_key)
        except GeminiError:
            raise
        except Exception as error:  # noqa: BLE001
            msg = str(error)
            self._logger.error("Gemini API error", error)
            if "API_KEY_INVALID" in msg or "400" in msg:
                raise GeminiError("Invalid Gemini API key. Please check your settings.")
            if "RATE_LIMIT" in msg or "429" in msg:
                raise GeminiError("Gemini rate limit exceeded. Please wait and retry.")
            if "SAFETY" in msg or "blocked" in msg:
                raise GeminiError("Image was blocked by Gemini safety filters.")
            if "quota" in msg or "QUOTA" in msg:
                raise GeminiError("Gemini API quota exceeded for today.")
            raise GeminiError(f"Gemini API error: {msg}")

    def _generate_sync(self, request: ImageGenerationRequest, api_key: str) -> ImageGenerationResponse:
        try:
            import google.generativeai as genai
        except ImportError:
            raise GeminiError("google-generativeai is not installed. Run: pip install google-generativeai")

        genai.configure(api_key=api_key)
        model = genai.GenerativeModel(
            model_name=request.model or "gemini-2.0-flash-exp",
            generation_config={"response_modalities": ["Text", "Image"]},
        )

        image_part = {"inline_data": {"mime_type": request.mime_type, "data": request.image_buffer}}
        result = model.generate_content([{"text": request.prompt}, image_part])

        result_buffer: Optional[bytes] = None
        result_mime = "image/png"
        for candidate in getattr(result, "candidates", None) or []:
            content = getattr(candidate, "content", None)
            for part in getattr(content, "parts", None) or []:
                inline = getattr(part, "inline_data", None)
                if inline and getattr(inline, "mime_type", "").startswith("image/"):
                    data = inline.data
                    result_buffer = data if isinstance(data, bytes) else base64.b64decode(data)
                    result_mime = inline.mime_type
                    break
            if result_buffer:
                break

        if not result_buffer:
            text = getattr(result, "text", "") or ""
            self._logger.warn("No image in Gemini response", {"textResponse": text[:200]})
            raise GeminiError(f"Gemini did not return an image. Response: {text[:200] or 'empty'}")

        tokens = None
        usage = getattr(result, "usage_metadata", None)
        if usage:
            tokens = getattr(usage, "total_token_count", None)

        self._logger.info("Gemini image generation successful", {
            "outputSize": len(result_buffer), "outputMimeType": result_mime, "tokensUsed": tokens,
        })
        return ImageGenerationResponse(image_buffer=result_buffer, mime_type=result_mime, tokens_used=tokens)

    async def validate_credentials(self) -> bool:
        api_key = os.environ.get("GEMINI_API_KEY")
        if not api_key:
            return False
        return await self.validate_api_key(api_key)

    async def validate_api_key(self, api_key: str) -> bool:
        def _check() -> bool:
            try:
                import google.generativeai as genai
                genai.configure(api_key=api_key)
                model = genai.GenerativeModel("gemini-1.5-flash")
                resp = model.generate_content('Say "ok" in one word.')
                return bool(getattr(resp, "text", "") )
            except Exception:
                return False
        return await asyncio.to_thread(_check)
